omnigibson/systems/macro_particle_system.py

```diff
@@ -1129,7 +932,344 @@ def _deserialize(cls, state):
+
+    @classmethod
+    def set_particle_velocities(cls, idx, lin_vel=None, ang_vel=None):
+        if lin_vel is None or ang_vel is None:
+            l_vel, a_vel = cls.get_particles_velocities()
+            lin_vel = l_vel if lin_vel is None else lin_vel
+            ang_vel = a_vel if ang_vel is None else ang_vel
+        cls.particles_view.set_velocities(np.concatenate([lin_vel, ang_vel]).reshape(1, -1), indices=np.array([idx]))
+
+    @classproperty
+    def particle_radius(cls):
+        return cls._particle_radius
+
+    @classproperty
+    def particle_contact_radius(cls):
+        # This is simply the normal radius
+        return cls.particle_radius
+
+    @classmethod
+    def generate_particles(
+            cls,
+            positions,
+            velocities=None,
+            angular_velocities=None,
+            orientations=None,
+            scales=None,
+            **kwargs,
+    ):
+        """
+        Generates new particles
+
+        Args:
+            positions (np.array): (n_particles, 3) shaped array specifying per-particle (x,y,z) positions
+            velocities (None or np.array): (n_particles, 3) shaped array specifying per-particle (x,y,z) velocities.
+                If not specified, all will be set to 0
+            angular_velocities (None or np.array): (n_particles, 3) shaped array specifying per-particle (ax,ay,az)
+                angular velocities. If not specified, all will be set to 0
+            orientations (None or np.array): (n_particles, 4) shaped array specifying per-particle (x,y,z,w) quaternion
+                orientations. If not specified, all will be sampled randomly
+            scales (None or np.array): (n_particles, 3) shaped array specifying per-particle (x,y,z) scales.
+                If not specified, will be uniformly randomly sampled from (cls.min_scale, cls.max_scale)
+            **kwargs (dict): Any additional keyword-specific arguments required by subclass implementation
+        """
+        # Call super first
+        super().generate_particles(
+            positions=positions,
+            velocities=velocities,
+            orientations=orientations,
+            scales=scales,
+            **kwargs,
+        )
+
+        # Grab pre-existing vels -- note that this already includes the newly included particles, so we will only
+        # keep the first (N - n_new) values
+        current_lin_vels, current_ang_vels = cls.get_particles_velocities()
+
+        # Update the tensors
+        n_particles = len(positions)
+        velocities = np.zeros((n_particles, 3)) if velocities is None else velocities
+        angular_velocities = np.zeros_like(velocities) if angular_velocities is None else angular_velocities
+
+        velocities = np.concatenate([current_lin_vels[:-n_particles], velocities], axis=0)
+        angular_velocities = np.concatenate([current_ang_vels[:-n_particles], angular_velocities], axis=0)
```

Comment:
So at this point, I assume the newly added particles have zero velocities by default?